### PR TITLE
Sensitivity analysis

### PR DESCRIPTION
**Pull request by sneha-desai**

### CODE REVIEW

File: vulcanai/models/metrics.py [hunk 1]
```diff
@@ -1012,3 +1014,85 @@ def cross_validate(network, data_loader, k, epochs,
...
+        the dataset and setting the feature for all examples to every unique value of that feature.
+
+        Parameters
+        ----------
+            network : BaseNetwork
+                Network descendant of BaseNetwork.
+            data_loader : DataLoader
+                A DataLoader object which contains all the data
+            filename: string
+                The name of the file to save the test results to
```
Comment (rfratila) [inline vulcanai/models/metrics.py]: I think you are missing the `features` input documentation

File: vulcanai/models/metrics.py [hunk 2]
```diff
@@ -1012,3 +1014,85 @@ def cross_validate(network, data_loader, k, epochs,
...
+
+        if network._num_classes is None or network._num_classes == 0:
+            raise ValueError('There\'s no classification layer')
+
+        dataset = data_loader.dataset
+
+        test_df = None
+        col_headers = None
+
+        for feature in range(len(features)):
```
Comment (rfratila) [inline vulcanai/models/metrics.py]: would `idx` (aka index) be a better name than feature if you are iterating the the index locations of the features instead of the features themselves? or something like feature_idx?

File: vulcanai/models/metrics.py [hunk 3]
```diff
@@ -1012,3 +1014,85 @@ def cross_validate(network, data_loader, k, epochs,
...
+            feature_val_list = dataset[:][0][:, feature].tolist()
+
+            unique_feature_values = list(np.unique(feature_val_list))
```
Comment (rfratila) [inline vulcanai/models/metrics.py]: potentially creating a `set` object of this list may be faster (instead of `np.unique`) to get the unique values? @nirtiac what do you say?

File: vulcanai/models/metrics.py [hunk 4]
```diff
@@ -1012,3 +1014,85 @@ def cross_validate(network, data_loader, k, epochs,
...
+
+            if len(unique_feature_values) > cutoff:
+                print("================================")
```
Comment (rfratila) [inline vulcanai/models/metrics.py]: Please use the logging `logger` functionality similar to what it looks like in the rest of metrics for consistency and greater logging flexibility

File: vulcanai/models/metrics.py [hunk 5]
```diff
@@ -1012,3 +1014,85 @@ def cross_validate(network, data_loader, k, epochs,
...
+                print("Cutting off number of features to be tested for feature "
+                      "{} from {} to {}".format(features[feature],
+                                                len(unique_feature_values),
+                                                cutoff))
+                print("================================")
+                unique_feature_values = unique_feature_values[0::int(
```
Comment (rfratila) [inline vulcanai/models/metrics.py]: what is this line doing?

File: vulcanai/models/metrics.py [hunk 6]
```diff
@@ -1012,3 +1014,85 @@ def cross_validate(network, data_loader, k, epochs,
...
+                        len(unique_feature_values) / cutoff)]
+            dataset_temp = copy.deepcopy(dataset)
```
Comment (rfratila) [inline vulcanai/models/metrics.py]: instead of doing a deep copy of `dataset`. Why not deep copy the data loader dataset? then you store one less instance of dataset in memory because you already treat it as immutable

File: vulcanai/models/metrics.py [hunk 7]
```diff
@@ -1012,3 +1014,85 @@ def cross_validate(network, data_loader, k, epochs,
...
+            for feature_value in range(len(unique_feature_values)):
+                dataset_temp[:][0][:, feature] = unique_feature_values[
+                    feature_value]
+                print("Testing value {} for feature {}".format(
```
Comment (rfratila) [inline vulcanai/models/metrics.py]: can we update this to use the `logger` functionality?

File: vulcanai/models/metrics.py [hunk 8]
```diff
@@ -1012,3 +1014,85 @@ def cross_validate(network, data_loader, k, epochs,
...
+                    unique_feature_values[feature_value], features[feature]))
+
+                targets = dataset[:][1].tolist()
+
+                raw_predictions = network.forward_pass(
+                    data_loader=data_loader)
+
+                predictions = self.transform_outputs(raw_predictions)
+
+                tp, _, fp, _ = Metrics.get_confusion_matrix_values(targets,
```
Comment (rfratila) [inline vulcanai/models/metrics.py]: why not use the `get_score` function to calculate whatever score you want?

File: vulcanai/run.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
+from torch.utils.data import DataLoader, Subset
```
Comment (rfratila) [inline vulcanai/run.py]: This looks like you made an example script for the sensitivity analysis, awesome! I think it is better suited to be in the `vulcanai/examples` folder with a more informative name, thanks!

File: vulcanai/run.py [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+from vulcanai.datasets import TabularDataset
+from vulcanai.models import ConvNet, DenseNet
+import sys
+sys.path.append('../')
+from models.metrics import conduct_sensitivity_analysis
+
+def my_test_dataset():
+    """Create a dataset by importing from the test csv"""
+    fpath = "/Users/snehadesai/Documents/Vulcan/vulcanai/tests/datasets/test_data/birthweight_reduced.csv"
```
Comment (rfratila) [inline vulcanai/run.py]: I would use a relative path depending on the actual package and not your installation of Vulcan

File: vulcanai/tests/models/conftest.py [hunk 11]
```diff
@@ -317,3 +317,23 @@ def multi_input_cnn_data(conv2D_net, conv3D_net, multi_input_dnn_data):
         (TensorDataset(torch.rand(size=[10, *conv3D_net.in_dim])), True, False),
         multi_input_dnn_data
     ])
+
+
+@pytest.fixture(scope="module")
+def dnn_class_single_value_2():
```
Comment (rfratila) [inline vulcanai/tests/models/conftest.py]: what does the 2 represent? please add a bit more information in the class documentation where this fixture is being used

File: vulcanai/tests/models/test_metrics.py [hunk 12]
```diff
@@ -330,3 +330,12 @@ def test_run_test_binary(self, metrics, cnn_class_binary):
                             'macro_npv', 'macro_f1', 'macro_auc']
 
         assert all(k in res_dict for k in required_metrics)
+
+    def test_sensitivity_analysis(self, metrics, dnn_class_single_value_2,
+                                  sensitivity_data_loader):
+        metrics.conduct_sensitivity_analysis(dnn_class_single_value_2,
```
Comment (rfratila) [inline vulcanai/tests/models/test_metrics.py]: perhaps you can look at the saved file that this function leaves after running to see if the expected outputs are correct. What we have here isn't necessarily enough. You can use the other test csv for it to be a more robust test

File: vulcanai/tests/models/conftest.py (hunk 11 above)
Comment (rfratila) [reply vulcanai/tests/models/conftest.py]: this may change on how you update the test to use the already embedded test csv file

File: vulcanai/run.py [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+            'dropout': 0.5,
+        },
+        num_classes=22
+    )
+
+
+
+def run_sensitivity_analysis(dnn_class_single_value_2,
+                              tabular_dataset_train_loader):
+    conduct_sensitivity_analysis(dnn_class_single_value_2,
```
Comment (rfratila) [inline vulcanai/run.py]: A thought: Maybe we should return some JSON object along with saving the csv in case we want to programmatically pass the results through an API? @nirtiac I'd also like to get your thoughts on this

File: vulcanai/models/metrics.py [hunk 14]
```diff
@@ -1012,3 +1014,85 @@ def cross_validate(network, data_loader, k, epochs,
...
+        if network._num_classes is None or network._num_classes == 0:
+            raise ValueError('There\'s no classification layer')
+
+        dataset = data_loader.dataset
+
+        test_df = None
+        col_headers = None
+
+        for feature in range(len(features)):
+            feature_val_list = dataset[:][0][:, feature].tolist()
```
Comment (nirtiac) [inline vulcanai/models/metrics.py]: so in this case the feature could simply be an index? that's a good solution. did you test it with both?

File: vulcanai/models/metrics.py (hunk 3 above)
Comment (nirtiac) [reply vulcanai/models/metrics.py]: yeah set is faster (https://bbengfort.github.io/observations/2017/05/02/python-unique-benchmark.html)

File: vulcanai/run.py (hunk 13 above)
Comment (nirtiac) [reply vulcanai/run.py]: yes I think we should. we need to standardize this interface though as all our metric functions return something else. so for now just return a python dictionary and save the file but it may be changed in the future.

File: vulcanai/run.py (hunk 9 above)
Comment (sneha-desai) [reply vulcanai/run.py]: I actually didn't mean to commit this file. It was just for debugging purposes since I couldn't interactively debug. Should I still add it in vulcanai/examples?

File: vulcanai/models/metrics.py (hunk 5 above)
Comment (sneha-desai) [reply vulcanai/models/metrics.py]: Since unique_feature_values are now sorted, it pulls cutoff # of features evenly spaced from the list

Comment (sneha-desai) [reply vulcanai/models/metrics.py]: I made a small mistake, will update in the next pr, it needs a math.ceil() instead of an int()

File: vulcanai/models/metrics.py (hunk 6 above)
Comment (sneha-desai) [reply vulcanai/models/metrics.py]: I don't get what you mean? Instead of this line I do data_loader_temp = copy.deepcopy(data_loader)? How will that save memory?

File: vulcanai/models/metrics.py (hunk 8 above)
Comment (sneha-desai) [reply vulcanai/models/metrics.py]: I'm only storing tp and fp though because those are the only ones that matter. The only way to get this would be through get_confusion_matrix(). I could do get_score(...,metrics='confusion_matrix') instead, but is that necessary?

File: vulcanai/run.py (hunk 9 above)
Comment (rfratila) [reply vulcanai/run.py]: yes if you can clean it up it's worth it

File: vulcanai/models/metrics.py (hunk 6 above)
Comment (rfratila) [reply vulcanai/models/metrics.py]: you set `dataset = dataloader.dataset` and then deep copy `dataset`. Why not just deep copy from `deepcopy(data loader.dataset)` instead? then you don't need to store the data in the local `dataset` variable

File: vulcanai/models/metrics.py (hunk 5 above)
Comment (rfratila) [reply vulcanai/models/metrics.py]: ok let's add a detailed comment about that. it's not immediately obvious what's going on

File: vulcanai/tests/models/test_metrics.py (hunk 12 above)
Comment (sneha-desai) [reply vulcanai/tests/models/test_metrics.py]: Hm I'm not really sure how to do this... The network weights I have are random for the test

Comment (rfratila) [reply vulcanai/tests/models/test_metrics.py]: It doesn't matter if the network weights are random, as long as you initialize them exactly the same way each time (hint: setting the seed) then you can have deterministic output of the model because your function will operate across the same values every-time in this scenario no? well then once it completes a test run, take a look at the file it saves, validate that it's been correctly calculated, and then hardcode the assertions to match exactly the values you find in the file. You can try to modify the function afterwards which would output a different value which would then fail your assertions. I hope laying it out like this makes sense

File: vulcanai/run.py (hunk 13 above)
Comment (sneha-desai) [reply vulcanai/run.py]: How should this dict be arranged? Right now, it returns a df with headers {Feature, Value, Number of examples classified as class 0, ... 1, all the way to n (num classes)}. List of dicts per example with these headers as keys? Or something else?

File: vulcanai/models/metrics.py (hunk 14 above)
Comment (sneha-desai) [reply vulcanai/models/metrics.py]: I haven't tested it with both, but it should work just as well. I think I should actually make the default value of features None, such that if it is None, I just create a new list that goes from 1 to size of num columns of data_loader.dataset

File: vulcanai/tests/models/test_metrics.py (hunk 12 above)
Comment (sneha-desai) [reply vulcanai/tests/models/test_metrics.py]: Yeah it does, thanks for the explanation, I'll update the assertions to reflect this.

File: vulcanai/run.py (hunk 9 above)
Comment (sneha-desai) [reply vulcanai/run.py]: Ok

File: vulcanai/models/metrics.py (hunk 6 above)
Comment (sneha-desai) [reply vulcanai/models/metrics.py]: I use dataset in other areas in the code (~ lines 1071, 1051). I could just update these to using dataloader.dataset instead of creating the copy beforehand if that's better?

Comment (rfratila) [reply vulcanai/models/metrics.py]: yes I think so - that's what other parts of metrics uses when you look around. @nirtiac just want to get a third opinion from you that this is ok?

File: vulcanai/run.py (hunk 13 above)
Comment (nirtiac) [reply vulcanai/run.py]: oh hmmmm, a dataframe may indeed be more appropriate then. just specify it in the function documentation.

File: vulcanai/models/metrics.py (hunk 14 above)
Comment (nirtiac) [reply vulcanai/models/metrics.py]: yeah I like that!

File: vulcanai/models/metrics.py (hunk 8 above)
Comment (nirtiac) [reply vulcanai/models/metrics.py]: yeah I think you're right Sneha it's good the way it is.

File: vulcanai/models/metrics.py [hunk 15]
```diff
@@ -1012,3 +1014,85 @@ def cross_validate(network, data_loader, k, epochs,
...
+                    col_headers = ["Feature", "Value"]
+                    for i in range(num_classes):
+                        col_headers.append("Number of examples classified as "
+                                           "class " + str(i))
+                    test_df = pd.DataFrame(columns=col_headers)
+
+                row_to_append = [features[feature],
+                                 unique_feature_values[feature_value]]
+
+                for i in range(len(tp)):
```
Comment (rfratila) [inline vulcanai/models/metrics.py]: @sneha-desai you can use list comprehension and `zip(tp, fp)` to iterate more cleanly through this list

File: vulcanai/models/metrics.py [hunk 16]
```diff
@@ -1012,3 +1014,85 @@ def cross_validate(network, data_loader, k, epochs,
...
+
+                predictions = self.transform_outputs(raw_predictions)
+
+                tp, _, fp, _ = Metrics.get_confusion_matrix_values(targets,
+                                                                   predictions)
+
+                if test_df is None and col_headers is None:
+                    num_classes = network._num_classes
+                    col_headers = ["Feature", "Value"]
+                    for i in range(num_classes):
```
Comment (rfratila) [inline vulcanai/models/metrics.py]: you can probably just use `network._num_classes`. no need to reassign a value you already have if you don't change it in-between

**Comment by sneha-desai** [comment]
-
